docs/source/federation-spec.mdx

```diff
@@ -0,0 +1,625 @@
+  }
+}
+```
+
+> We strongly recommend _against_ performing composition dynamically within the graph router, because a composition error can cause unexpected downtime. Nevertheless, supporting this use case is still a requirement for subgraph libraries.
+
+### Differences from built-in introspection
+
+The "enhanced" introspection query above differs from [the GraphQL spec's built-in introspection query](https://spec.graphql.org/October2021/#sec-Schema-Introspection) in the following ways:
+
+- The returned schema representation is a string instead of a `__Schema` object.
+- The returned schema string includes all uses of federation-specific directives, such as `@key`.
+    - The built-in introspection query's response does _not_ include the uses of any directives.
+    - The graph router requires these federation-specific directives to perform composition successfully.
+-  If a subgraph server "disables introspection", the enhanced introspection query is still available.
+
+> Note that the `_service` field is _not_ included in the composed supergraph schema. For security reasons, it's intended solely for use by the graph router.
+
+### Required resolvers
+
+To support the enhanced introspection query, a subgraph service must define resolvers for the following fields:
+
+- `Query._service`
+- `_Service.sdl`
+
+`Query._service` returns a `_Service` object, which in turn has a single field, `sdl` (short for schema definition language). The `sdl` field returns a string representation of the subgraph's schema.
+
+The returned `sdl` string has the following requirements:
+
+* It must **include** all uses of all federation-specific directives, such as `@key`.
+    * For a list of all federation-specific subgraph directives, see the [glossary](#glossary-of-schema-additions).
+* It must **omit** all automatically added definitions from the [subgraph schema specification](#subgraph-schema-specification), such as `Query._service` and `_Service.sdl`!
+
+For example, consider this Federation 2 subgraph schema:
+
+```graphql title="schema.graphql"
+extend schema
+  @link(url: "https://specs.apollo.dev/federation/v2.0",
+        import: ["@key"])
+
+type Query {
+  me: User
+}
+
+type User @key(fields: "id") {
+  id: ID!
+}
+```
+
+The value returned for the `sdl` field should include all of this information, including directives (excessive whitespace can be removed).
+
+## Generating stub types
+
+Each subgraph in a supergraph should be able to run successfully in isolation, regardless of whether any _other_ subgraphs are running. Many subgraphs run successfully in isolation with no additional logic required. However, subgraph fields _can_ depend on types and fields that are defined in _other_ subgraphs. Whenever this is the case, a subgraph needs to define its own **stubs** for these external types and fields.
+
+To indicate a dependency on another subgraph's type or field, a subgraph schema field uses either the `@requires` or `@provides` directive:
```

Comment:
maybe we should also add an example for `@provides`?

Reply:
I'm going to leave it to one example in the federated directives article, just because the listings for `@requires` and `@provides` are both very close by and provide their own examples